File: leetcode/657.scala
object Solution {
    def judgeCircle(moves: String): Boolean = {
        moves.foldLeft((0, 0))( (pos, c: Char) => {
            c match {
                case 'U' => (pos._1, pos._2 + 1)
                case 'D' => (pos._1, pos._2 - 1)
                case 'R' => (pos._1 + 1, pos._2)
                case 'L' => (pos._1 - 1, pos._2)
                case _   => pos
            }
        }) == (0, 0)
    }
}

File: leetcode/657.rs
impl Solution {
    pub fn judge_circle(moves: String) -> bool {
        (0, 0) == moves.chars()
            .fold((0, 0), |(x, y), c| {
                match c {
                    'U' => (x, y + 1),
                    'D' => (x, y - 1),
                    'R' => (x + 1, y),
                    'L' => (x - 1, y),
                    _ => (x, y),
                }
             })
    }
}
